File: binding/include/charts/libqdatetimeaxis.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qdatetimeaxis.html
pub const qdatetimeaxis = struct {

    /// New constructs a new QDateTimeAxis object.
    pub fn New() QtC.QDateTimeAxis {
        return qtc.QDateTimeAxis_new();
    }


    /// New2 constructs a new QDateTimeAxis object.
    pub fn New2(parent: ?*anyopaque) QtC.QDateTimeAxis {
        return qtc.QDateTimeAxis_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QDateTimeAxis_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qdatetimeaxis.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Type(self: ?*anyopaque, ) i32 {
        return qtc.QDateTimeAxis_Type(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnType(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QDateTimeAxis_OnType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseType(self: ?*anyopaque, ) i32 {
        return qtc.QDateTimeAxis_QBaseType(@ptrCast(self));
    }

    pub fn SetMin(self: ?*anyopaque, min: QtC.QDateTime) void {
        qtc.QDateTimeAxis_SetMin(@ptrCast(self), @ptrCast(min));
    }

    pub fn Min(self: ?*anyopaque, ) QtC.QDateTime {
        return qtc.QDateTimeAxis_Min(@ptrCast(self));
    }

    pub fn SetMax(self: ?*anyopaque, max: QtC.QDateTime) void {
        qtc.QDateTimeAxis_SetMax(@ptrCast(self), @ptrCast(max));
    }

    pub fn Max(self: ?*anyopaque, ) QtC.QDateTime {
        return qtc.QDateTimeAxis_Max(@ptrCast(self));
    }

    pub fn SetRange(self: ?*anyopaque, min: QtC.QDateTime, max: QtC.QDateTime) void {
        qtc.QDateTimeAxis_SetRange(@ptrCast(self), @ptrCast(min), @ptrCast(max));
    }

    pub fn SetFormat(self: ?*anyopaque, format: []const u8) void {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
qtc.QDateTimeAxis_SetFormat(@ptrCast(self), format_str);
    }

    pub fn Format(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QDateTimeAxis_Format(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qdatetimeaxis.Format: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetTickCount(self: ?*anyopaque, count: i32) void {
        qtc.QDateTimeAxis_SetTickCount(@ptrCast(self), count);
    }

    pub fn TickCount(self: ?*anyopaque, ) i32 {
        return qtc.QDateTimeAxis_TickCount(@ptrCast(self));
    }

    pub fn MinChanged(self: ?*anyopaque, min: QtC.QDateTime) void {
        qtc.QDateTimeAxis_MinChanged(@ptrCast(self), @ptrCast(min));
    }

    pub fn OnMinChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, QtC.QDateTime) callconv(.c) void) void {
        qtc.QDateTimeAxis_Connect_MinChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MaxChanged(self: ?*anyopaque, max: QtC.QDateTime) void {
        qtc.QDateTimeAxis_MaxChanged(@ptrCast(self), @ptrCast(max));
    }

    pub fn OnMaxChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, QtC.QDateTime) callconv(.c) void) void {
        qtc.QDateTimeAxis_Connect_MaxChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn RangeChanged(self: ?*anyopaque, min: QtC.QDateTime, max: QtC.QDateTime) void {
        qtc.QDateTimeAxis_RangeChanged(@ptrCast(self), @ptrCast(min), @ptrCast(max));
    }

    pub fn OnRangeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, QtC.QDateTime, QtC.QDateTime) callconv(.c) void) void {
        qtc.QDateTimeAxis_Connect_RangeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn FormatChanged(self: ?*anyopaque, format: []const u8) void {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
qtc.QDateTimeAxis_FormatChanged(@ptrCast(self), format_str);
    }

    pub fn OnFormatChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QDateTimeAxis_Connect_FormatChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn TickCountChanged(self: ?*anyopaque, tick: i32) void {
        qtc.QDateTimeAxis_TickCountChanged(@ptrCast(self), tick);
    }

    pub fn OnTickCountChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QDateTimeAxis_Connect_TickCountChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QDateTimeAxis_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qdatetimeaxis.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QDateTimeAxis_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qdatetimeaxis.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QDateTimeAxis_Delete(@ptrCast(self));
    }
};

File: binding/src/charts/libqdatetimeaxis.cpp
#include <QAbstractAxis>
#include <QDateTime>
#include <QDateTimeAxis>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qdatetimeaxis.h>
#include "libqdatetimeaxis.h"
#include "libqdatetimeaxis.hxx"

QDateTimeAxis* QDateTimeAxis_new() {
	 return new VirtualQDateTimeAxis();
}

QDateTimeAxis* QDateTimeAxis_new2(QObject* parent) {
	 return new VirtualQDateTimeAxis(parent);
}

libqt_string QDateTimeAxis_Tr(const char* s) {
	QString _ret = QDateTimeAxis::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QDateTimeAxis_Type(const QDateTimeAxis* self) {
	return self->type();
}

void QDateTimeAxis_SetMin(QDateTimeAxis* self, QDateTime* min) {
	self->setMin(*min);
}

QDateTime* QDateTimeAxis_Min(const QDateTimeAxis* self) {
	return new QDateTime(self->min());
}

void QDateTimeAxis_SetMax(QDateTimeAxis* self, QDateTime* max) {
	self->setMax(*max);
}

QDateTime* QDateTimeAxis_Max(const QDateTimeAxis* self) {
	return new QDateTime(self->max());
}

void QDateTimeAxis_SetRange(QDateTimeAxis* self, QDateTime* min, QDateTime* max) {
	self->setRange(*min, *max);
}

void QDateTimeAxis_SetFormat(QDateTimeAxis* self, libqt_string format) {
	self->setFormat(QString::fromUtf8(format.data, format.len));
}

libqt_string QDateTimeAxis_Format(const QDateTimeAxis* self) {
	QString _ret = self->format();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QDateTimeAxis_SetTickCount(QDateTimeAxis* self, int count) {
	self->setTickCount(count);
}

int QDateTimeAxis_TickCount(const QDateTimeAxis* self) {
	return self->tickCount();
}

void QDateTimeAxis_MinChanged(QDateTimeAxis* self, QDateTime* min) {
	self->minChanged(*min);
}

void QDateTimeAxis_Connect_MinChanged(QDateTimeAxis* self, intptr_t slot) {
    void (*slotFunc)(QDateTimeAxis*, QDateTime*) = reinterpret_cast<void (*)(QDateTimeAxis*, QDateTime*)>(slot);
    QDateTimeAxis::connect(self, &QDateTimeAxis::minChanged, [self, slotFunc](QDateTime min) {
	slotFunc(self, min);
    });
}

void QDateTimeAxis_MaxChanged(QDateTimeAxis* self, QDateTime* max) {
	self->maxChanged(*max);
}

void QDateTimeAxis_Connect_MaxChanged(QDateTimeAxis* self, intptr_t slot) {
    void (*slotFunc)(QDateTimeAxis*, QDateTime*) = reinterpret_cast<void (*)(QDateTimeAxis*, QDateTime*)>(slot);
    QDateTimeAxis::connect(self, &QDateTimeAxis::maxChanged, [self, slotFunc](QDateTime max) {
	slotFunc(self, max);
    });
}

void QDateTimeAxis_RangeChanged(QDateTimeAxis* self, QDateTime* min, QDateTime* max) {
	self->rangeChanged(*min, *max);
}

void QDateTimeAxis_Connect_RangeChanged(QDateTimeAxis* self, intptr_t slot) {
    void (*slotFunc)(QDateTimeAxis*, QDateTime*, QDateTime*) = reinterpret_cast<void (*)(QDateTimeAxis*, QDateTime*, QDateTime*)>(slot);
    QDateTimeAxis::connect(self, &QDateTimeAxis::rangeChanged, [self, slotFunc](QDateTime min, QDateTime max) {
	slotFunc(self, min, max);
    });
}

void QDateTimeAxis_FormatChanged(QDateTimeAxis* self, libqt_string format) {
	self->formatChanged(QString::fromUtf8(format.data, format.len));
}

void QDateTimeAxis_Connect_FormatChanged(QDateTimeAxis* self, intptr_t slot) {
    void (*slotFunc)(QDateTimeAxis*, const char*) = reinterpret_cast<void (*)(QDateTimeAxis*, const char*)>(slot);
    QDateTimeAxis::connect(self, &QDateTimeAxis::formatChanged, [self, slotFunc](QString format) {
	slotFunc(self, format);
    });
}

void QDateTimeAxis_TickCountChanged(QDateTimeAxis* self, int tick) {
	self->tickCountChanged(tick);
}

void QDateTimeAxis_Connect_TickCountChanged(QDateTimeAxis* self, intptr_t slot) {
    void (*slotFunc)(QDateTimeAxis*, int) = reinterpret_cast<void (*)(QDateTimeAxis*, int)>(slot);
    QDateTimeAxis::connect(self, &QDateTimeAxis::tickCountChanged, [self, slotFunc](int tick) {
	slotFunc(self, tick);
    });
}

libqt_string QDateTimeAxis_Tr2(const char* s, const char* c) {
	QString _ret = QDateTimeAxis::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QDateTimeAxis_Tr3(const char* s, const char* c, int n) {
	QString _ret = QDateTimeAxis::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
int QDateTimeAxis_QBaseType(const QDateTimeAxis* self) {
	auto* vqdatetimeaxis = dynamic_cast<const VirtualQDateTimeAxis*>(self);
	if (vqdatetimeaxis && vqdatetimeaxis->isVirtualQDateTimeAxis) {
vqdatetimeaxis->setQDateTimeAxis_Type_IsBase(true);
	return vqdatetimeaxis->type();
}
}

// Auxiliary method to allow providing re-implementation
void QDateTimeAxis_OnType(const QDateTimeAxis* self, intptr_t slot) {
	auto* vqdatetimeaxis = dynamic_cast<const VirtualQDateTimeAxis*>(self);
	if (vqdatetimeaxis && vqdatetimeaxis->isVirtualQDateTimeAxis) {
vqdatetimeaxis->setQDateTimeAxis_Type_Callback(reinterpret_cast<VirtualQDateTimeAxis::QDateTimeAxis_Type_Callback>(slot));
}
}

void QDateTimeAxis_Delete(QDateTimeAxis* self) {
    delete self;
}

